fairlearn/metrics/_group_metric_set.py

```diff
@@ -0,0 +1,169 @@
+    BINARY_CLASSIFICATION_METRICS[GROUP_RECALL_SCORE] = group_recall_score
+    BINARY_CLASSIFICATION_METRICS[GROUP_ROC_AUC_SCORE] = group_roc_auc_score
+    BINARY_CLASSIFICATION_METRICS[GROUP_SELECTION_RATE] = group_selection_rate
+    BINARY_CLASSIFICATION_METRICS[GROUP_SPECIFICITY_SCORE] = group_specificity_score
+
+    REGRESSION_METRICS = {}
+    REGRESSION_METRICS[GROUP_BALANCED_ROOT_MEAN_SQUARED_ERROR] = group_balanced_root_mean_squared_error  # noqa:E501
+    REGRESSION_METRICS[GROUP_MAX_ERROR] = group_max_error
+    REGRESSION_METRICS[GROUP_MEAN_ABSOLUTE_ERROR] = group_mean_absolute_error
+    REGRESSION_METRICS[GROUP_MEAN_OVERPREDICTION] = group_mean_overprediction
+    REGRESSION_METRICS[GROUP_MEAN_PREDICTION] = group_mean_prediction
+    REGRESSION_METRICS[GROUP_MEAN_SQUARED_ERROR] = group_mean_squared_error
+    REGRESSION_METRICS[GROUP_MEAN_SQUARED_LOG_ERROR] = group_mean_squared_log_error
+    REGRESSION_METRICS[GROUP_MEAN_UNDERPREDICTION] = group_mean_underprediction
+    REGRESSION_METRICS[GROUP_MEDIAN_ABSOLUTE_ERROR] = group_median_absolute_error
+    REGRESSION_METRICS[GROUP_R2_SCORE] = group_r2_score
+    REGRESSION_METRICS[GROUP_ROOT_MEAN_SQUARED_ERROR] = group_root_mean_squared_error
+    REGRESSION_METRICS[GROUP_ZERO_ONE_LOSS] = group_zero_one_loss
+
+    def __init__(self):
+        self._model_type = None
+        self._y_true = None
+        self._y_pred = None
+        self._groups = None
+        self._group_names = None
+        self._metrics = None
+
+    @property
+    def model_type(self):
+        """Return the type of model being described."""
+        return self._model_type
+
+    @model_type.setter
+    def model_type(self, value):
+        if value not in GroupMetricSet._allowed_model_types:
+            msg_format = "model_type '{0}' not in {1}"
+            msg = msg_format.format(value, sorted(list(GroupMetricSet._allowed_model_types)))
+            raise ValueError(msg)
+        self._model_type = value
+
+    @property
+    def y_true(self):
+        """Return the array of y_true values."""
+        return self._y_true
+
+    @y_true.setter
+    def y_true(self, value):
+        self._y_true = _convert_to_ndarray_1d(value)
+
+    @property
+    def y_pred(self):
+        """Return the array of y_pred values."""
+        return self._y_pred
+
+    @y_pred.setter
+    def y_pred(self, value):
+        self._y_pred = _convert_to_ndarray_1d(value)
+
+    @property
+    def groups(self):
+        """Return the array of group values."""
+        return self._groups
+
+    @groups.setter
+    def groups(self, value):
+        self._groups = _convert_to_ndarray_1d(value)
+
+    @property
+    def group_names(self):
+        """Return the group_names dictionary."""
+        return self._group_names
+
+    @group_names.setter
+    def group_names(self, value):
+        key_types = set(type(k) for k in value.keys())
+        if key_types != {int}:
+            raise ValueError("Keys not integers")
```

Comment:
can we have full sentences in these error messages?
e.g. "The type of the argument `value` for `group_names` needs to be `int`. The provided type was ...."

Reply:
Exception messages all improved